src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -2150,47 +2151,67 @@ private unsafe void DispatchToAllListeners(int eventId, EventWrittenEventArgs ev
 
         private unsafe void WriteEventString(EventLevel level, long keywords, string msgString)
         {
+#if FEATURE_MANAGED_ETW || FEATURE_PERFTRACING
+            bool allAreNull = true;
 #if FEATURE_MANAGED_ETW
-            if (m_etwProvider != null)
+            allAreNull &= (m_etwProvider == null);
+#endif // FEATURE_MANAGED_ETW
+#if FEATURE_PERFTRACING
+            allAreNull &= (m_eventPipeProvider == null);
+#endif // FEATURE_PERFTRACING
+            if (allAreNull)
+            {
+                return;
+            }
+
+            const string EventName = "EventSourceMessage";
+            if (SelfDescribingEvents)
             {
-                const string EventName = "EventSourceMessage";
-                if (SelfDescribingEvents)
+                EventSourceOptions opt = new EventSourceOptions
                 {
-                    EventSourceOptions opt = new EventSourceOptions
-                    {
-                        Keywords = (EventKeywords)unchecked(keywords),
-                        Level = level
-                    };
-                    var msg = new { message = msgString };
-                    var tlet = new TraceLoggingEventTypes(EventName, EventTags.None, new Type[] { msg.GetType() });
-                    WriteMultiMergeInner(EventName, ref opt, tlet, null, null, msg);
+                    Keywords = (EventKeywords)unchecked(keywords),
+                    Level = level
+                };
+                var msg = new { message = msgString };
+                var tlet = new TraceLoggingEventTypes(EventName, EventTags.None, new Type[] { msg.GetType() });
+                WriteMultiMergeInner(EventName, ref opt, tlet, null, null, msg);
```

Comment:
Reviewing the code in WriteMultiMergeInner I think it has the same level caching bug that the new code in EventWriteTransfer has. I expect that if you call WriteEventString() multiple times with different levels every message will be logged as with the same level as the first.

Reply:
I don't think that's an issue for the reason I mentioned above. WriteEventString is only for error messages and should always be LogAlways, it's saying there is a problem and the EventSource is unusable

Reply:
I filed #35393 to track the general issue.